pilot/pkg/model/service.go

```diff
@@ -481,8 +483,22 @@ type IstioEndpoint struct {
 	// Labels points to the workload or deployment labels.
 	Labels labels.Instance
 
-	// Address is the address of the endpoint, using envoy proto.
-	Address string
+	// Addresses is the address of the endpoint, using envoy proto:
+	// https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/endpoint/v3/endpoint_components.proto#config-endpoint-v3-endpoint-additionaladdress
+	// This filed can support multiple addresses for an Dual Stack endpoint, especially for an endpoint which contains both ipv4 or ipv6 addresses.
+	// There should be some constraints below:
+	// 1. Each address of the endpoint must have the same metadata.
+	// 2. The function Key() of IstioEndpoint returns the first IP address of this field in string format.
+	// 3. The IP address of field `address` in Envoy Endpoint is equal to the first address of this field.
+	// When the additional_addresses field is populated for EDS in Envoy configuration, there would be a Happy Eyeballs algorithm to
+	// instantiate for the Envoy Endpoint, the first attempt connecting to the IP address in the `address` field of Envoy Endpoint.
+	// Thereafter it will interleave IP addresses in the `additional_addresses` field based on IP version, as described in rfc8305,
+	// and attempt connections with them with a delay of 300ms each. The first connection to succeed will be used.
```

Comment:
I am not convinced happy eyeballs is the right approach in Istio. We should have more discussion here outside of this PR.

Reply:
This results in very different behavior than K8s without Istio IIUC.

Reply:
Hi @howardjohn, sure, we can have more discussion about this. And EDS in Envoy provides this feature and `containerd` supports the Happy Eyeballs algorithm as well. However, we can decide to use it or not, if not, we'd better to have some mechanism or rules to make sure that the Dual Stack is supported.
BTW, I think the task of K8s is filling the IP addresses to service, pod and endpoint/slice, etc, but K8s CNI and container runtime really do the connected actions.

Reply:
I am not so sure about that? In ~every CNI, the IP family chosen is based on the application. I don't know of any CNI that would get an IPv4 requests and send to an IPv6 backend. I am not even sure it complies with k8s specification for Service...